### PR TITLE
update katib-ui rockcraft.yaml

### LINKED ISSUE #32: `katib-ui-rock` is based on wrong Dockerfile
It turns out we have ROCKed the wrong dockerfile for `katib-ui-rock`.

The katib-ui component we 've charmed is the until recently (release `0.15`) called `new-ui` component in the upstream. We can see this in the way they [produced images](https://github.com/kubeflow/katib/blob/v0.15.0/.github/workflows/publish-core-images.yaml#L27-L28). Looking at the`0.15` release branch, this is the [dockerfile](https://github.com/kubeflow/katib/blob/v0.15.0/cmd/ui/v1beta1/Dockerfile) we rocked while the image we use in our charm comes from [this one](https://github.com/kubeflow/katib/blob/v0.15.0/cmd/new-ui/v1beta1/Dockerfile). With release `0.16`, [they removed the old ui](https://github.com/kubeflow/katib/pull/2179) and thus renamed previous `new-ui` component to `ui`. 

Summarizing, we need to update the ROCK to be based on [this dockerfile](https://github.com/kubeflow/katib/blob/v0.16.0-rc.1/cmd/ui/v1beta1/Dockerfile) (`ui` from release `0.16-rc.1`).

### PR DESCRIPTION
**Pull request by ca-scribner**
Previously, an older version of the katib-ui had been rocked.  This commit updates the rock to the current katib-ui, replicating the build defined by https://github.com/kubeflow/katib/blob/v0.16.0/cmd/ui/v1beta1/Dockerfile

Note:
* the version of the frontend-lib pulled here is that specified by the katib repo in the [COMMIT](https://github.com/kubeflow/katib/blob/v0.16.0/pkg/ui/v1beta1/frontend/COMMIT) file of the upstream repo
* the GOARCH is omitted from building here because upstream also omits this.  This might be something to revisit later, or might be something determined in different ways in rocks anyway?

Closes #32 

#### Testing instructions

Do:
```bash
VERSION=version_in_rockcraft_file
rockcraft pack
skopeo --insecure-policy copy oci-archive:katib-ui_${VERSION}_amd64.rock docker-daemon:katib-ui:${VERSION}
docker save katib-ui:${VERSION} > katib_ui_${VERSION}.tar
microk8s ctr image import katib_ui_${VERSION}.tar
```
Then do:
* `juju deploy katib-ui --resource oci-image=katib-ui:${VERSION}` using a charm version includes [this fix to the katib-ui charm](https://github.com/canonical/katib-operators/pull/160)
* browse to `KATIB_UI_SERVICE_IP:8080/katib/` and you should see the katib UI

### CODE REVIEW

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [inline katib-ui-rock/rockcraft.yaml:68]: nit but would it make sense to copy those the other way around in order make sure `node_modules/kubeflow` is intact? (you can always just resolve :smile: )

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [inline katib-ui-rock/rockcraft.yaml:34]: I remember not having this creating some prompts during `npm ci` (I think), but do we need this?

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [inline katib-ui-rock/rockcraft.yaml:81]: Why 1.20? Maybe it would make sense to document how did you decide on the go version since this is not straight forward form looking at the Dockerfile (so we can be cautious during updates it in the future)

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [inline katib-ui-rock/rockcraft.yaml:37]: What about `GOARCH`?

Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:37]: Hmm looking from how they build their images in CI ([here](https://github.com/kubeflow/katib/blob/v0.16.0/.github/workflows/publish-core-images.yaml)), I see they do not define this anywhere.Is that why you skipped it too?

**Review by orfeas-k — COMMENTED** [review]
Thank you for the PR @ca-scribner, great and structured work. I left some comments that are up for discussion. Didn't deploy and test the ROCK yet but will do tomorrow.

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:6]: Yeah if canonical/bundle-kubeflow#747 closes quick then we should edit this version, but otherwise maybe we merge this PR as is and edit this again if needed when canonical/bundle-kubeflow#747 is finalised?  

I'm open either way, but to me its better to update this rock now with a best guess at version, then modify the version later as needed

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:26]: good catch - fixed the `source-commit` and broke to two lines

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:28]: Good points re this comment, it doesn't need to explain the backstory.  I'm removing the extra note, as I think the comment above it covers things.  wdyt?

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:29]: I think the end of your comment captures my worry about the beginning :D

It does feel odd that a user has to keep `frontend` and `frontend-lib`'s git checkouts in sync, especially since the `katib/katib` repo checked out by `frontend` includes that `COMMIT` file that specifies the commit hash we want to use from `kubeflow/kubeflow` when building `frontend-lib`.  We could remove that human-sync by taking the commit hash from the `COMMIT` file like you describe.

Where it gets awkward I think is that we'd have to pass it inside the rockcraft parts (since we can't dynamically write to `source-commit` in the rockcraft.yaml).  What we could do is, in `frontend-lib`, in a override-build do `commit=$(cat /root/parts/frontend/src/path/to/COMMIT); git checkout kubeflow/kubeflow $commit; ...`.  That would remove the human-sync effort, but make a more complicated rock file.  wdyt?

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:31]: I'll delete them.  We cannot uncomment `source-depth` (because it is incompatible with `source-commit`, which checks out a single commit), but leaving this here at all really isn't needed.

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:34]: I honestly dont know :)  I see it in some but not all of our rocks, and some dockerfiles, which makes me think it isn't actually necessary?

Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:34]: I don't mind removing, I just don't really know the consequences.  I do think if we omit it, the analytics default to `true`, which I think means our built app might be sending telemetry to someone which I think we don't want

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:39]: Yeah this is more a me-pattern and I can remove it if needed.

By default, the `source: https://somewhere/in/github` directive pulls the source to `/root/parts/PARTNAME/src` and then copies that source to `/root/parts/PARTNAME/build`.  Most of our rocks then just `cd` to the dir that has the source and build there.

Upstream dockerfiles do this differently.  They don't pull the whole repo, but instead just `COPY` the files needed for the build (kinda like what I'm doing in this rockfile).  This felt more natural imo because it means I only build with the things that actually matter.

I'm paranoid (mainly because I don't know this software, and don't know if leaving extra files around could affect the build) so I went with the "copy only what I need" approach like the upstream dockerfile had.  That's why I do a `rm -rf ./`.  But it is probably overkill.  

wdyt?

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:68]: Sorry I don't understand what you mean.  Could you give an example?

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:81]: I have no idea - tbh I'm not even sure why I picked 1.20.  The upstream dockerfile is not specific either

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:37]: Yeah I omitted it just because they do, but I don't know if that is appropriate.

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml]: Yep agreed for the comment, looks good now. WDYT though about adding more details for this in PR or issue?

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:32]: Yep they might be sent to Angular project. Good point actually, let's keep it then

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:29]: Oh I didn't realise this is possible actually. Hmm, this feels a better approach (less error prone updates) but isn't weird/not recommended to access files from another part in terms `rockcraft` best practices? I 'm OK with the file as is for the reasons we both mentioned, so I 'll leave this up to you.

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:66]:
```suggestion
      cp -r $CRAFT_PART_SRC/pkg/ui/v1beta1/frontend/* .
      cp -r $CRAFT_STAGE/frontend-lib/dist/kubeflow ./node_modules/kubeflow
```
That's my suggestion.

In order to explain my comment more, `$CRAFT_PART_SRC/pkg/ui/v1beta1/frontend/*` could include in theory a `.../frontend/node_modules/kubeflow/` folder. We copy everything and `node_modules` folder is being created during `npm ci` which has already run. Thus it would overwrite the previously copied `node_modules/kubeflow`. It struck me mainly because [upstream](https://github.com/kubeflow/katib/blob/v0.16.0/cmd/ui/v1beta1/Dockerfile#L32-L33) makes those copies with a different order. In practice that won't happen since katib-ui doesn't install any `kubeflow` module/package. But still, following what upstream does would protect us from this edge case scenario.

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:37]: Yes, I think it's nice that we do it the way upstream does it. So IIUC `rm -rf ./` deletes the source code from `/root/parts/PARTNAME/build` and then copies only what we need right? If that's the case, I think we can keep it. To your concerns, having the repo's parent directories around wouldn't play any role since `npm build` uses specific files.

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml]: Got it [here](https://github.com/kubeflow/katib/blob/v0.16.0/go.mod#L3). (looked at the [CI](https://github.com/kubeflow/katib/blob/v0.16.0/.github/workflows/test-go.yaml#L47)). We can keep 1.20 with a comment pointing to that file

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:37]: I think it's fine since we build the image ourselves. I mean, even if we add it, that won't change something since in their case, I guess the CI could at some point provide that variable in order to build images for another Architecture. In our case, I guess building for a different architecture would require to modify other parts of rockcraft.yaml.

That being said, I think we should document either inside the rockcraft.yaml or the issue responsible for it some of this information, so one won't have to look at the review comments to understand why this is omitted.

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml]: oh yeah I misunderstood the comment before.  I'll add that now

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:29]: Yeah it feels anti-rockcraft, but also handy.  But maybe too complex for the effort it saves? 

I'll leave it as is, but we could discuss with the team in future.  sound good?

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:37]: Yes you're right about what is happening (remove source from `.../build`, then copy the files needed)

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:66]: ohhh yeah totally agree good catch! I've updated that now

File: katib-ui-rock/rockcraft.yaml
Comment (ca-scribner) [reply katib-ui-rock/rockcraft.yaml:37]: k i'll add a note to the issue

File: katib-ui-rock/rockcraft.yaml
Comment (orfeas-k) [reply katib-ui-rock/rockcraft.yaml:29]: Sure, we could. We 'll figure this out for good in case someone else (apart from us) updates this for 1.9 release. Added `source-commit` in the [ROCK upgrade section](https://docs.google.com/document/d/1sCTnY3T4b6FEv5jmHlyB2PYC2ZGCY4tAB4cKTk-R7FY/edit#heading=h.u5ruz84494xq) in the spec

**Review by orfeas-k — APPROVED** [review]
Thank you @ca-scribner, tested it on a cluster and I can verify this works! Great job

**Pull request merged**
